File: src/claude_skills/catalog.py
from __future__ import annotations

import json
import os
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

import yaml  # type: ignore[import-untyped]

from claude_skills.models import Catalog, CatalogMetadata, Skill


class CatalogBuilder:
    BASE = ".claude/skills"

    def __init__(self, root: Optional[Path] = None):
        self.root = Path(root or os.getcwd())
        self.skills_dir = self.root / self.BASE

    def scan(self) -> list[Skill]:
        skills: list[Skill] = []
        seen: set[str] = set()

        for sk_path in sorted(self.skills_dir.rglob("SKILL.md")):
            meta = self._parse_frontmatter(sk_path)
            if not meta or "name" not in meta:
                continue

            name = meta["name"]
            if name in seen:
                continue
            seen.add(name)

            ru_path = sk_path.parent / "SKILL.ru.md"
            has_ru = ru_path.exists()

            skill = Skill(
                name=name,
                description=meta.get("description", ""),
                category=meta.get("category", ""),
                tags=self._parse_list(meta.get("tags", [])),
                models=self._parse_list(meta.get("models", [])),
                version=str(meta.get("version", "1.0.0")),
                path=sk_path,
                languages=["en", "ru"] if has_ru else ["en"],
                has_ru=has_ru,
                created=str(meta.get("created", "")),
                updated=str(meta.get("updated", "")),
            )
            skills.append(skill)

        return skills

    def build_catalog(self) -> Catalog:
        skills = self.scan()
        domains = sorted(set(s.category for s in skills))
        now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

        metadata = CatalogMetadata(
            schema_version="3.0",
            generated_at=now,
            total_skills=len(skills),
            total_ru=sum(1 for s in skills if s.has_ru),
            domains=domains,
            bilingual=True,
        )
        return Catalog(metadata=metadata, skills=skills)

    def to_json(self, catalog: Catalog, path: Optional[Path] = None) -> str:
        data = {
            "metadata": {
                "schema_version": catalog.metadata.schema_version,
                "generated_at": catalog.metadata.generated_at,
                "total_skills": catalog.metadata.total_skills,
                "total_ru": catalog.metadata.total_ru,
                "domains": catalog.metadata.domains,
                "bilingual": catalog.metadata.bilingual,
            },
            "skills": [
                {
                    "name": s.name,
                    "description": s.description,
                    "category": s.category,
                    "tags": s.tags,
                    "models": s.models,
                    "version": s.version,
                    "path": str(s.path.as_posix()),
                    "languages": s.languages,
                    "has_ru": s.has_ru,
                    "created": s.created or "",
                    "updated": s.updated or "",
                }
                for s in catalog.skills
            ],
        }
        text = json.dumps(data, indent=2, ensure_ascii=False)
        if path:
            path.write_text(text, encoding="utf-8")
        return text

    def _parse_frontmatter(self, filepath: Path) -> Optional[dict]:
        try:
            content = filepath.read_text(encoding="utf-8")
            if not content.startswith("---"):
                return None
            end = content.find("---", 3)
            if end < 0:
                return None
            front = content[3:end].strip()
            return yaml.safe_load(front) or {}
        except (yaml.YAMLError, OSError, UnicodeDecodeError):
            return None

    def _parse_list(self, value) -> list[str]:
        if isinstance(value, list):
            return [str(v).strip() for v in value if v]
        if isinstance(value, str):
            cleaned = value.strip().strip("[]").strip()
            if not cleaned:
                return []
            return [v.strip().strip("\"'") for v in re.split(r"[\s,]+", cleaned) if v.strip()]
        return []

    @staticmethod
    def from_json(path: Path) -> Catalog:
        data = json.loads(path.read_text(encoding="utf-8"))
        meta = data.get("metadata", {})
        metadata = CatalogMetadata(
            schema_version=meta.get("schema_version", "3.0"),
            generated_at=meta.get("generated_at", ""),
            total_skills=meta.get("total_skills", 0),
            total_ru=meta.get("total_ru", 0),
            domains=meta.get("domains", []),
            bilingual=meta.get("bilingual", True),
        )
        skills = [
            Skill(
                name=s["name"],
                description=s.get("description", ""),
                category=s.get("category", ""),
                tags=s.get("tags", []),
                models=s.get("models", []),
                version=str(s.get("version", "1.0.0")),
                path=Path(s.get("path", "")),
                languages=s.get("languages", ["en"]),
                has_ru=s.get("has_ru", False),
                created=s.get("created", ""),
                updated=s.get("updated", ""),
            )
            for s in data.get("skills", [])
        ]
        return Catalog(metadata=metadata, skills=skills)

File: web/src/lib/catalog.ts
export interface Skill {
  name: string
  description: string
  category: string
  tags: string[]
  models: string[]
  version: string
  path: string
  languages: string[]
  has_ru: boolean
  created: string
  updated: string
}

export interface CatalogData {
  metadata: {
    schema_version: string
    generated_at: string
    total_skills: number
    total_ru: number
    domains: string[]
    bilingual: boolean
  }
  skills: Skill[]
}

export async function loadCatalog(): Promise<CatalogData> {
  if (typeof window === "undefined") {
    const fs = await import("fs")
    const path = await import("path")
    const filePath = path.join(process.cwd(), "public", "data", "catalog.json")
    const raw = fs.readFileSync(filePath, "utf-8")
    const data: CatalogData = JSON.parse(raw)
    return normalizeCatalog(data)
  }
  const res = await fetch("/data/catalog.json")
  if (!res.ok) throw new Error("Failed to load catalog")
  const data: CatalogData = await res.json()
  return normalizeCatalog(data)
}

function normalizeCatalog(data: CatalogData): CatalogData {
  for (const skill of data.skills) {
    if (typeof skill.tags === "string") {
      skill.tags = (skill.tags as string).split(",").map((t: string) => t.trim()).filter(Boolean)
    }
    if (typeof skill.models === "string") {
      skill.models = (skill.models as string).split(",").map((t: string) => t.trim()).filter(Boolean)
    }
  }
  return data
}

export function getCategories(catalog: CatalogData): string[] {
  return catalog.metadata.domains.sort()
}

export function getSkillBySlug(catalog: CatalogData, slug: string): Skill | undefined {
  return catalog.skills.find((s) => s.name === slug)
}

export function getSkillsByCategory(catalog: CatalogData, category: string): Skill[] {
  return catalog.skills.filter((s) => s.category === category)
}

export function getRelatedSkills(catalog: CatalogData, skill: Skill, limit = 6): Skill[] {
  const tagSet = new Set(skill.tags.map((t) => t.toLowerCase()))
  const scored = catalog.skills
    .filter((s) => s.name !== skill.name)
    .map((s) => {
      const overlap = s.tags.filter((t) => tagSet.has(t.toLowerCase())).length
      const sameCat = s.category === skill.category ? 1 : 0
      return { skill: s, score: overlap * 3 + sameCat * 2 }
    })
    .sort((a, b) => b.score - a.score)
  return scored.slice(0, limit).map((s) => s.skill)
}
